### ISSUE TITLE
Using user name / git user name inside code snippets

### ISSUE THREAD

**Issue by kachkaev** [Microsoft/vscode#64599]
I'd like to add the following code snippet and share it with the whole team:

```json
{
	"Comment todo": {
		"prefix": "//",
		"body": [
			"// TODO (${MY_NAME}) [${CURRENT_YEAR}-${CURRENT_MONTH}-${CURRENT_DATE}]: $1"
		]
	}
}
```

This should produce `// TODO (Alexander Kachkaev) [2018-12-07]: ...` on my computer and `// TODO (John Smith) [2018-12-07]: ...` on John Smith's.

Looking at [VSCode snippet docs](https://code.visualstudio.com/docs/editor/userdefinedsnippets#_variables) and [VSCode variable reference](https://code.visualstudio.com/docs/editor/variables-reference) does not suggest that there is any variable like `MY_NAME`. Could it be added? The value can either be taken from the OS user name or from the git user name. WDYT?

**Comment by iveselin** [comment]
Exactly what I am searching for right now, any updates or plans?

**Comment by aetheric8** [comment]
access to the environment variables like in tasks with ${env:USER} or even include shell commands ${cmd:`echo $USER`} would be very useful

**Comment by GlenNicholls** [comment]
Is there any workaround to add an author with a specific command? Maybe an environment variable, Git username, or os username?

**Comment by azra1l** [comment]
i'd also like to use git user values (specifically author / email) to include those in my changelog snippet.
everything else i need is already in place and that's so awesome =)

**Comment by claytonjhamilton** [comment]
Yes, I would love to have my git username imported from the global config file into my VS Code snippets. That would be a time saver!!

**Comment by aminomancer** [comment]
I agree with azra, the important thing is to at least let snippets access settings so extension developers can implement it themselves without burdening vscode's developers much. But I think there are a few more variables that whould ideally be exposed to snippets, namely the OS and git usernames. As an example, I made a simple snippet that creates a greasemonkey style metadata block:

```
// ==UserScript==
// @name           {filename}
// @namespace      https://github.com/{username}/{folder}/{filename}
// @author         {username}
// @version        {version}
// @homepage       https://github.com/{username}/{folder}
// @description    {MM/DD/YY}
// ==/UserScript==
```
Which works nicely but as you can probably already tell, there is no username variable. I'm able to insert filename and date directly, and I can get folder from TM_DIRECTORY with regex. But username is just something the user is gonna have to input directly, every time, even when vscode should definitely be aware of what the remote repo address is.

Best case scenario, when the file is in a repo, vscode should expose some info about the repo to snippets, particularly the full address of the file, (not its local file path but its path relative to the repo) the repo name, and the name of the host and the owner if remote. Shouldn't be too hard to at least make it work for git and mercurial hosts

**Comment by Gruntfuggly** [comment]
If environment variables could be inserted into snippets, there would at least be a work around. Feels like it shouldn't be a difficult update?

**Comment by TheDevelolper** [comment]
Perhaps it could get the author info from git config?

**Comment by lucas0headshot** [comment]
Still unsolved?

**Comment by RedCMD** [comment]
I'm sure a PR would be welcome

the starting file to modify [snippetVariables.ts](https://github.com/microsoft/vscode/blob/main/src/vs/editor/contrib/snippet/browser/snippetVariables.ts)

and there is also [snippetsFile.ts](https://github.com/microsoft/vscode/blob/main/src/vs/workbench/contrib/snippets/browser/snippetsFile.ts)

**Comment by dmitrivMS** [comment]
Some thoughts: It's not hard to expose env variables, however I am not sure that is the right approach as it would rely on VS Code process having the right env variables for whatever workspace is loaded which is probably not always the case. Ideally user info should come from Git and/or auth session, which is more complicated since it would have to be injected into snippets which is in core from extensions. If we go that route, the first step may be to add an API for extensions to contribute snippets variables, then git can supply one.

FWIW, here's the PR with env variables - I am curious if it really solves the needs described above?
https://github.com/microsoft/vscode/pull/315156

Also related: https://github.com/microsoft/vscode/issues/63534

**Comment by TheDevelolper** [comment]
I think you're right, extensibility is really the way to go. For now, this is an interesting work around. 

I think there should be two separate variables for the OS user and Git user this will give flexibility. 

What about fallback values too? 

```jsonc
{
  "Comment todo": {
    "prefix": "//",
    "body": [
        "// TODO (${OS_NAME ?? GIT_NAME})
    ]
  }
}
```

Perhaps the extensibility part could be satisfied using some sort of notation for calculated variables. Perhaps supporting a JSX like format alongside what's there already (you just change the file extension if you want to use it). 

This will give the user a lot more repo specific flexibility without having to create extensions for snippets. 

I appreciate that ideas are cheap implementation costs. The great thing about that though, is that we can afford to have a lot of ideas 🙂.
